### DISCUSSION TITLE
Discount code applicable/validation from Cart endpoint

### DISCUSSION THREAD

**Discussion by icrustandi** [Shopify/storefront-api-feedback#22] [Help]
Hi,

As we're trying to move off from using the checkout object to implement our cart on storefront API using the Cart endpoint, I noticed that we don't have access to whether or not an applied discount code is applicable on existing cart items. This used to be an accessable field on Checkout, but I don't see the applicable boolean anywhere on the Cart docs.

https://shopify.dev/api/storefront/reference/checkouts/cartdiscountcode
vs
https://shopify.dev/api/storefront/reference/discounts/discountcodeapplication

Is there any way to get the applicable information from the cart?

**Comment by icrustandi** [comment]
I'm also noticing that there is no validation whatsoever on the cartDiscountCodesUpdate mutation. On the checkout object, there are three states a discount code payload can have:
1. completely invalid code - nothing created in discountCodes
2. Valid code but not applicable
3. Valid and applicable

https://shopify.dev/api/storefront/reference/checkouts/cartdiscountcodesupdate

Will this feature be implemented on cart?

**Reply by morganmccunn** [reply]
Thanks for the feedback @icrustandi ! This is definitely something we'll need to be implementing on the cart. I will follow up once we have this in place.

**Comment by morganmccunn** [comment]
Hey @icrustandi we've made a couple changes here:
* Added the `applicable` field for CartDiscountCode to verify whether a discount count was applicable to the current cart.
* Added validation for `discount_code` 

For the validation, we are intentionally returning the same user error regardless of whether the code is (1) completely invalid or (2) valid but not applicable in an effort to prevent from brute force attacks on guessing discount codes.

**Comment by icrustandi** [comment]
@morganmccunn I see that applicable has been added to the discountCode field here but there's some weird behavior here still. I've laid out a few scenarios - some of these have to do with checkout not syncing back to the original cart, so it may be two different issues compounding on each other.


My setup is a BOGO code setup, let's just call the discount code BOGO, where the customer is able to get one product variant for free if they have two product variants in the cart (lesser of the two value is free).
Scenario 1:
1.  I have the two qualifying product variants in the cart
2. Apply applicable code BOGO
3. storefront API responds back with the correct discountCode allocations, discounted amounts are correct, and applicable is correctly set to true
4. I refresh my page to re-fetch my cart object
5. storefront API responds back with a cart that has what looks to be the correct discountCode allocations but applicable is set to false when it should be set to true

Expected behavior - re-fetching the cart object should display persistent discount code allocations and applicable should remain true

Scenario 2:
1. Continuing off of scenario 1 actions... remove previously applied BOGO discount code using the `cartDiscountCodesUpdate` mutation (payload of empty discountCodes)
2. discountCodes field in cart is empty but discountAllocations on qualifying item from previous BOGO deal still remains in the cart, previously qualifying item is incorrectly discounted as free

Expected behavior - Removing promo code should also remove discountAllocations on the individual previously qualifying line item

Scenario 3:
1. Continuing off of the flawed state of the cart.. go to checkout using checkoutUrl
2. Checkout has no promo codes in cart
3. Go back to same cart - cart is still in the same state as the end of scenario 2 (empty discountCodes but discountAllocations on previously qualifying item still exists

Expected behavior - if checkout has no promo codes applied, discountAllocations should not exist when going back to cart

Scenario 4:
1. Empty cart
2. Add two qualifying items without applying discount code
3. Proceed to checkout using checkoutUrl
4. On checkout, apply BOGO code
5. On checkout, the qualifying item is correctly discounted as free
6. Refresh checkout - checkout still correctly shows discounted item
7. Go back to cart - cart responds with two qualifying items without the BOGO code applied, line items are not discounted
8. Proceed to checkout using webUrl
9. On checkout, no promo codes are applied, list item prices are shown

Expected behavior - if promo codes are applied on checkout, the cart should also have the same state of having the promocode applied

**Reply by haeky** [reply]
Hey @icrustandi,

We've made some changes to handle most of the scenarios you've described above.

1. applicable field should be queryable and keeps the correct value whether the discount code was applicable to the current cart.
2. Added/removed discount should carry over properly to the checkout page.

As for syncing between the checkout and the cart, this is a larger undertaking but we're working on this. We'll keep you posted 👍

**Reply by icrustandi** [reply]
@haeky great, it looks like the only scenario that remains is scenario 4. 

It seems like the current behavior for cart - checkout sync only works one way? Cart -> checkout works fine but checkout -> cart doesn't. Is the expectation that there will be a working two-way sync cart <-> checkout once the cart API moves out of the unstable channel?

**Reply by Muneefm** [reply]
@haeky is `discountCodes` in cartDiscountCodesUpdate case sensitive?  for example i created a discount code "sale", but when i enter "Sale" instead it returns `applicable` false

**Reply by jullite** [reply]
@Muneefm Yes the discountCodes is case sensitive
![image](https://user-images.githubusercontent.com/86504730/211499070-6bb51821-9c5b-43d5-b87d-caf06ecc61c0.png)

**Comment by jullite** [comment]
@haeky As we're trying to utilize `cartDiscountCodesUpdate` to implement code applying instead of `checkoutDiscountCodeApplyV2` on our cart , I found I can't get the error message when I apply an inapplicable code to cart like we use `checkoutDiscountCodeApplyV2` mutation. We need that message to tell our users why that code can not been applied. Maybe instead of returning an applicable state for that code, throwing the error on `userErrors` will be more helpful.

**Comment by lennart-devhelden** [comment]
Any updates on how one can get the error of why a discount Code couldn't be applied to the cart? This is necessary for giving a good UX for the users. Right now we can only show can be applied or not, but this is not enough for us. The Shopify Web store does this already, so why not applying it to Storefront? We have large Customers using Shopify, and they want better messages for the users (e.g.  "You need to do this or that for the code to be applied") for their apps using Storefront.
